platform/pwl.c

```diff
@@ -1218,74 +1274,35 @@ static void
 registry_global_remove (void *data, struct wl_registry *registry, uint32_t name)
 {
     PwlDisplay *display = data;
-    PwlOutput *item, *tmp;
-    wl_list_for_each_safe (item, tmp, &display->outputs, link) {
-        if (item->output_name == name) {
+    PwlOutput *pwl_output, *tmp_output;
+    wl_list_for_each_safe (pwl_output, tmp_output, &display->outputs, link) {
+        if (pwl_output->output_name == name) {
             g_debug ("Output #%"PRIi32" @ %p: Removed.",
-                     item->output_name, item->output);
-            g_clear_pointer (&item->output, wl_output_release);
-            wl_list_remove (&item->link);
-            g_slice_free (PwlOutput, item);
+                     pwl_output->output_name, pwl_output->output);
+            g_clear_pointer (&pwl_output->output, wl_output_release);
+            wl_list_remove (&pwl_output->link);
+            g_slice_free (PwlOutput, pwl_output);
             break;
         }
     }
+    PwlSeat *pwl_seat, *tmp_seat;
+    wl_list_for_each_safe (pwl_seat, tmp_seat, &display->seats, link) {
+        g_debug ("Seat @ %p: Removed.",
+                 pwl_seat->seat);
+        g_clear_pointer (&pwl_seat->seat, wl_seat_release);
+        wl_list_remove (&pwl_seat->link);
+        g_slice_free (PwlSeat, pwl_seat);
+    }
```

Comment:
This loop is removing _all_ the `PwlSeat` instances from the list! What is worse: it's done unconditionally, so if an output is removed, all the seats will be destroyed as well.

You need to compare the `name` integer to check whether the object being destroyed is one of the elements contained in the list of seats. Check how the same is done right above when iterating the list of outputs. You may need to add a `seat_name` integer to `PwlSeat` to store the object identifier when the seat is created, so later on it can be used here for the comparison.

Reply:
Oh, now i understand the reason of keep the `name` value as part of the struct. Thanks for the review.

Code simplified and encapsulated in the `pwl_seat_destroy()`